xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -228,79 +239,243 @@ private static LdsUpdate processClientSideListener(Listener listener, boolean pa
       }
     }
 
-    // Parse RDS info.
-    if (hcm.hasRouteConfig()) {
-      // Found inlined route_config. Parse it to find the cluster_name.
+    // Parse inlined RouteConfiguration or RDS.
+    if (proto.hasRouteConfig()) {
       List<VirtualHost> virtualHosts = new ArrayList<>();
       for (io.envoyproxy.envoy.config.route.v3.VirtualHost virtualHostProto
-          : hcm.getRouteConfig().getVirtualHostsList()) {
-        StructOrError<VirtualHost> virtualHost = parseVirtualHost(virtualHostProto, parseFilter);
+          : proto.getRouteConfig().getVirtualHostsList()) {
+        StructOrError<VirtualHost> virtualHost =
+            parseVirtualHost(virtualHostProto, parseHttpFilter);
         if (virtualHost.getErrorDetail() != null) {
-          throw new ResourceInvalidException("HttpConnectionManager contains invalid virtual host: "
-              + virtualHost.getErrorDetail());
+          throw new ResourceInvalidException(
+              "HttpConnectionManager contains invalid virtual host: "
+                  + virtualHost.getErrorDetail());
         }
         virtualHosts.add(virtualHost.getStruct());
       }
-      return new LdsUpdate(maxStreamDuration, virtualHosts, filterChain);
+      return io.grpc.xds.HttpConnectionManager.forVirtualHosts(
+          maxStreamDuration, virtualHosts, filterConfigs);
     }
-
-    if (hcm.hasRds()) {
-      // Found RDS.
-      Rds rds = hcm.getRds();
+    if (proto.hasRds()) {
+      Rds rds = proto.getRds();
       if (!rds.hasConfigSource()) {
-        throw new ResourceInvalidException("HttpConnectionManager missing config_source for RDS.");
+        throw new ResourceInvalidException(
+            "HttpConnectionManager contains invalid RDS: missing config_source");
       }
       if (!rds.getConfigSource().hasAds()) {
         throw new ResourceInvalidException(
-            "HttpConnectionManager ConfigSource for RDS does not specify ADS.");
+            "HttpConnectionManager contains invalid RDS: must specify ADS");
       }
-      return new LdsUpdate(maxStreamDuration, rds.getRouteConfigName(), filterChain);
+      return io.grpc.xds.HttpConnectionManager.forRdsName(
+          maxStreamDuration, rds.getRouteConfigName(), filterConfigs);
     }
-
     throw new ResourceInvalidException(
-        "HttpConnectionManager neither has inlined route_config nor RDS.");
-  }
-
-  private LdsUpdate processServerSideListener(Listener listener)
-      throws ResourceInvalidException {
-    StructOrError<EnvoyServerProtoData.Listener> convertedListener =
-        parseServerSideListener(listener, tlsContextManager);
-    if (convertedListener.getErrorDetail() != null) {
-      throw new ResourceInvalidException(convertedListener.getErrorDetail());
-    }
-    return new LdsUpdate(convertedListener.getStruct());
+        "HttpConnectionManager neither has inlined route_config nor RDS");
   }
 
   @VisibleForTesting
```

Comment:
Comment is for line 492: this particular code (the error "HttpFilter ... is not optional and has no typed config") is not tested anywhere. Before this unification there were tests which have been removed.

Reply:
See `parseHttpFilter_unsupportedButOptional()` and `parseHttpFilter_unsupportedAndRequired()`.